design-proposals/2023-02-08-code-organisation.md

```diff
@@ -0,0 +1,137 @@
+
+Testing should be done by `CTest` if the platform allows (may not be possible for all integration tests on mobile devices).
+
+### Release cycle
+
+The core should have an independent release cycle. At this point we imagine SDKs to be much more actively developed. This also allows bigger refactoring to take place without affecting SDKs (e.g. Metal migration).
+
+## SDKs and other bindings
+
+SDKs, language bindings and platform bindings should provide utilities to integrate MapLibre better to a specific platform. They should not add any core mapping functionality but they could add integration utilities that can not be made platform independent, e.g.
+
+- widgets
+- gestures support on toucscreen devices
+- location services (GPS, location from the mobile network)
+
+### Organisation
+
+MapLibre core could be provided as a submodule, but it is not necessary. For mobile platforms this may or may not be the case, depending on the level of the integration, for example:
+
+- on Android if Kotlin/Java is mostly used, then distributing the core separately as a dynamic library is completely reasonable
+- on iOS where static building may be much more beneficial due to size the core will probably be built together with the SDK
+
+We leave the SDK organisation up to the specific platform developers. It should probably follow conventions of the platform to simplify distribution and usage.
+
+The build system is also not explicitly defined. It would be strongly preferred that the core is built by `CMake` (or at least the necessary build scripts are not duplicated in a SDK repository/folder but generated by `CMake`).
+
+### Scope and feature-parity
+
+Different SDKs and bindings **are not expected to have feature-parity**. They can/will live in separate repositories and will be developed independently.
+
+When a feature is added to the core it may not be added to all SDKs at the same time.
+
+## Repository split - yes or no?
+
+This proposal does not touch much on repository splitting. If SDKs continue to live in the main repository, they should be put under `sdk/<platform>`. The following primary SDKs should be created and supported based on the currently available code:
+
+- Android SDK
+- Apple SDK (common between all Apple platforms)
+- Qt SDK
+- Node.js Binaries and SDK
+- GLFW SDK
+
+Navigation SDKs could then be merged with respective platform SDK.
+
+### Repository naming
+
+- `maplibre-native-core` or `maplibre-native-renderer` for the core functionality.
+- SDKs:
+  - `maplibre-native-sdk-apple` (iOS and macOS, potentially merged with `maplibre-navigation-ios`)
+  - `maplibre-native-sdk-android` (Android, potentially merged with `maplibre-navigation-android`)
+  - `maplibre-native-sdk-qt` (Qt bindings, should probably be merged with `qt-geoservices-maplibre-gl`)
+  - `maplibre-native-sdk-js`
+
+Alternatively SDKs could be without the explicit `sdk` in the name.
```

Comment:
I am definitely in favor of dropping the `sdk` part.

Reply:
Same

Reply:
OK